applications/serial_lte_modem/src/gnss/slm_at_gnss.c

```diff
@@ -0,0 +1,681 @@
+	}
+}
+#endif /* CONFIG_SLM_AGPS */
+
+static void on_gnss_evt_agps_req(void)
+{
+#if defined(CONFIG_SLM_AGPS)
+	int err1 = nrf_modem_gnss_read((void *)&agps_data, sizeof(agps_data),
+					NRF_MODEM_GNSS_DATA_AGPS_REQ);
+	if (err1) {
+		LOG_ERR("Failed to read GNSS PVT data, error %d", err1);
+		return;
+	}
+	k_work_submit(&agps_req);
+#endif /* CONFIG_SLM_AGPS */
+
+#if defined(CONFIG_SLM_PGPS)
+	int err2;
+
+	err2 = nrf_cloud_pgps_request_all();
+	if (err2) {
+		LOG_ERR("Failed to request P-GPS data: %d", err2);
+	}
+	err2 = nrf_cloud_pgps_notify_prediction();
+	if (err2) {
+		LOG_ERR("Failed to request notification of prediction: %d", err2);
+	}
+#endif /* CONFIG_SLM_PGPS */
+}
+
+/* NOTE this event handler runs in interrupt context */
+static void gnss_event_handler(int event)
+{
+	switch (event) {
+	case NRF_MODEM_GNSS_EVT_PVT:
+		LOG_INF("NRF_MODEM_GNSS_EVT_PVT");
+		break;
+	case NRF_MODEM_GNSS_EVT_FIX: {
+		struct nrf_modem_gnss_pvt_data_frame pvt;
+		int err;
+
+		LOG_INF("NRF_MODEM_GNSS_EVT_FIX");
+		if (ttft_start != 0) {
+			LOG_INF("TTFF %ds", (int)k_uptime_delta(&ttft_start)/1000);
+			ttft_start = 0;
+		}
+		err = nrf_modem_gnss_read((void *)&pvt, sizeof(pvt), NRF_MODEM_GNSS_DATA_PVT);
+		if (err) {
+			LOG_ERR("Failed to read GNSS PVT data, error %d", err);
+			break;
+		}
+		sprintf(rsp_buf,
+			"\r\n#XGPS: %lf,%lf,%f,%f,%f,%f,\"%04u-%02u-%02u %02u:%02u:%02u\"\r\n",
+			pvt.latitude, pvt.longitude, pvt.altitude,
+			pvt.accuracy, pvt.speed, pvt.heading,
+			pvt.datetime.year, pvt.datetime.month, pvt.datetime.day,
+			pvt.datetime.hour, pvt.datetime.minute, pvt.datetime.seconds);
+		rsp_send(rsp_buf, strlen(rsp_buf));
+#if defined(CONFIG_SLM_PGPS)
+		struct tm gps_time = {
+			.tm_year = pvt.datetime.year - 1900,
+			.tm_mon  = pvt.datetime.month - 1,
+			.tm_mday = pvt.datetime.day,
+			.tm_hour = pvt.datetime.hour,
+			.tm_min  = pvt.datetime.minute,
+			.tm_sec  = pvt.datetime.seconds,
+		};
+		date_time_set(&gps_time);
+		nrf_cloud_pgps_set_location(pvt.latitude, pvt.longitude);
+#endif /* CONFIG_SLM_PGPS */
+	} break;
+	case NRF_MODEM_GNSS_EVT_NMEA:
+		LOG_INF("NRF_MODEM_GNSS_EVT_NMEA");
+		break;
+	case NRF_MODEM_GNSS_EVT_AGPS_REQ:
+		LOG_INF("NRF_MODEM_GNSS_EVT_AGPS_REQ");
+		on_gnss_evt_agps_req();
+		break;
+	case NRF_MODEM_GNSS_EVT_BLOCKED:
+		LOG_WRN("NRF_MODEM_GNSS_EVT_BLOCKED");
+		break;
+	case NRF_MODEM_GNSS_EVT_UNBLOCKED:
+		LOG_INF("NRF_MODEM_GNSS_EVT_UNBLOCKED");
+		break;
+	case NRF_MODEM_GNSS_EVT_PERIODIC_WAKEUP:
+		LOG_DBG("NRF_MODEM_GNSS_EVT_PERIODIC_WAKEUP");
+		break;
+	case NRF_MODEM_GNSS_EVT_SLEEP_AFTER_TIMEOUT:
+		LOG_DBG("NRF_MODEM_GNSS_EVT_SLEEP_AFTER_TIMEOUT");
+		break;
+	case NRF_MODEM_GNSS_EVT_SLEEP_AFTER_FIX:
+		LOG_DBG("NRF_MODEM_GNSS_EVT_SLEEP_AFTER_FIX");
+		break;
+	case NRF_MODEM_GNSS_EVT_REF_ALT_EXPIRED:
+		LOG_DBG("NRF_MODEM_GNSS_EVT_REF_ALT_EXPIRED");
+		break;
+	default:
+		break;
+	}
+}
+
+#if defined(NRF_CLOUD_ACCESS)
+static void on_cloud_evt_ready(void)
```

Comment:
if the user has multiple services (pgps, agps, cell_pos) enabled, they will be executed here, even if they didn't request it.  is that correct?  is that the desired behavior?

Reply:
Yes, this is desired behavior as any of the CONFIG would need nRF Cloud access.
In this function we use CONFIG to control what's needs to be done upon CLOUD_EVT_READY.

Reply:
if the user enables PGPS, AGPS, and cell_pos in their config and then sends the `AT#XCELLPOS` AT command, this will execute PGPS, AGPS and cell_pos.
this is what you want? 
as a user i would expect only cell_pos to be executed.

Reply:
You are right. The expected use cases are GPS only, GPS + AGPS, GPS + PGPS, CELL_POS only. In case of CELL_POS, GPS is enabled but not started. I'll add condition checks to make sure of above scenarios.

Reply:
@junqingzou I misread your comment and thought you mentioned AGPS + PGPS as a use case.  This misunderstanding led me to make my comments in my review.  My mistake.